tests/test_locks.py

```diff
@@ -0,0 +1,50 @@
+from loguru import logger
+
+
+class CyclicReference:
+    """A minimal cyclic reference.
+
+    Cyclical references are garbage collected using the generational collector rather than
+    via reference counting. This is important here, because the generational collector runs
+    periodically, meaning that it is hard to predict when the stack will be overtaken by a
+    garbage collection process - but it will almost always be when allocating memory of some
+    kind.
+
+    When this object is garbage-collected, a log will be emitted.
+    """
+    def __init__(self, _other: "CyclicReference" = None):
+        self.other = _other or CyclicReference(_other=self)
+
+    def __del__(self):
+        logger.info("tearing down")
+
+
+def test_no_deadlock_on_generational_garbage_collection():
+    """Regression test for https://github.com/Delgan/loguru/issues/712
+
+    Assert that deadlocks do not occur when a cyclic isolate containing log output in
+    finalizers is collected by generational GC, during the output of another log message.
+    """
+
+    # GIVEN a sink which assigns some memory
+    output = []
+
+    def sink(message):
+        # This simulates assigning some memory as part of log handling. Eventually this will
+        # trigger the generational garbage collector to take over here.
+        # You could replace it with `json.dumps(message.record)` to get the same effect.
+        _ = [dict() for _ in range(20)]
+
+        # Actually write the message somewhere
+        output.append(message)
+
+    logger.add(sink, colorize=False)
+
+    # WHEN there are cyclic isolates in memory which log on GC
+    # AND logs are produced long enough to trigger generational GC
+    for _ in range(1000):
```

Comment:
This number is somewhat arbitrary unfortunately. The generational GC will trigger after a certain amount of memory is allocated. Possibly this depends on the amount of memory available? On my machine `range(10)` did not reproduce the issue, but `range(100)` did. I went for `range(1000)` to be safe. Its still quite quick - 130ms on my machine for the entire test.